mm2src/lp_swap/trade_preimage.rs

```diff
@@ -289,6 +298,55 @@ impl From<CoinFindError> for TradePreimageRpcError {
     }
 }
 
+impl From<MakerOrderBuildError> for TradePreimageRpcError {
+    fn from(e: MakerOrderBuildError) -> Self {
+        match e {
+            MakerOrderBuildError::BaseEqualRel => TradePreimageRpcError::BaseEqualRel,
+            MakerOrderBuildError::MaxBaseVolTooLow { actual, threshold } => TradePreimageRpcError::VolumeIsTooSmall {
+                volume: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            MakerOrderBuildError::PriceTooLow { actual, threshold } => TradePreimageRpcError::PriceTooLow {
+                price: actual.to_decimal(),
+                threshold: threshold.to_decimal(),
+            },
+            MakerOrderBuildError::RelVolTooLow { actual, threshold } => TradePreimageRpcError::VolumeIsTooSmall {
```

Comment:
Unfortunately, it is a bit complicated to separate the `TradePreimageRpcError::VolumeIsTooSmall` to `TradePreimageRpcError::BaseVolumeIsTooSmall` and `TradePreimageRpcError::RelVolumeIsTooSmall`.

Reply:
I don't get why that is. Can't we just have two different errors for `BaseVolumeIsTooSmall `and `RelVolumeIsTooSmall` here instead of one
https://github.com/KomodoPlatform/atomicDEX-API/blob/30d962c392293a470f97ace63f079953f9db297c/mm2src/lp_swap/trade_preimage.rs#L213
Just to be known what is meant by the `threshold` because for rel volume it also depends on the price.

Reply:
>  Just to be known what is meant by the threshold because for rel volume it also depends on the price.

Great point! `threshold` is a minimum rel coin volume. The threshold price can be calculated by:
`threshold_price = threshold / base_coin_volume`, if `method` is `buy`

Reply:
`CheckBalanceError::from_trade_preimage_error` is called also at the start of the taker swap. But at this point, there is no information about which coin was `base` and which is `rel`. So as we discussed with @yurii-khi, it would be convenient for both MarketMaker and GUI to handle an error like
`VolumeTooLow { volume: BigDecimal, threshold: BigDecimal, coin: String }`. It will allow us to determine if it was a base coin volume or a rel coin volume.

Reply:
> `CheckBalanceError::from_trade_preimage_error` is called also at the start of the taker swap. But at this point, there is no information about which coin was `base` and which is `rel`. So as we discussed with @yurii-khi, it would be convenient for both MarketMaker and GUI to handle an error like
> `VolumeTooLow { volume: BigDecimal, threshold: BigDecimal, coin: String }`. It will allow us to determine if it was a base coin volume or a rel coin volume.

Is it possible to have some json example in case of error ? so i can start adding the data structure in cpp in my case

Reply:
@shamardy, could you please recheck the last changes?

Reply:
> Is it possible to have some json example in case of error ? so i can start adding the data structure in cpp in my case

I'm working on docs right now. Please wait until I finished :)